### DISCUSSION TITLE
Python Text annotation with PyMuPDF

### DISCUSSION THREAD

**Discussion by testingdlrna** [pymupdf/PyMuPDF#873] [Q&A] [answered]
I'm using PyMuPDF for annotating some text in . pdf document by using:

`import fitz 
import re 

def data_(text): 
        
        annotation_text = r"(amet)"
        for line in text:
            if re.search(annotation_text, line, re.IGNORECASE): 
                search = re.search(annotation_text, line, re.IGNORECASE) 
                yield search.group(1) 

    def includeannotation(path_included): 
        
        document = fitz.open(path_included) 
        
        
        for page in document: 
            page.wrap_contents() 
            obs = data_(page.getText("text") .split('\n'))
            #print (obs)
            for data in obs: 
                catchs = page.searchFor(data) 
                [page.addRedactAnnot(catchs, fontsize=11, fill = (0, 0, 0)) for catch in catchs] 
            page.apply_redactions() 
        doc.save('annotation.pdf') 
        print("end - created") 

path_included = '/content/document.pdf'

save_document=includeannotation(path_included)`

The source .pdf document contains the text:
<img width="751" alt="YtLwm" src="https://user-images.githubusercontent.com/64439766/106490192-de844280-64b5-11eb-9704-6cc089cf03ab.png">

By applying the above mentioned code, I can include the annotation for the text "amet" obtain the following result:

<img width="494" alt="lvcvu" src="https://user-images.githubusercontent.com/64439766/106490237-e8a64100-64b5-11eb-8cd4-f53188e2af59.png">

And the result seems to be in line with the expection, but you can see that the library has included the annotation in black (for "amet") also deleting the word in the line after, but not with the black annotation. And in fact it looks like a restyling problem.

How can I avoid such problem?

**Comment by JorjMcKie** [comment, marked as answer]
This is a case of text lines which overlap each other in their line height.
When applying redactions, MuPF remove all characters whose bbox has a **_non-empty overlap_** with a redaction rectangle.To avoid this, globally set the character bbox height to fontsize height by executing ``fitz.TOOLS.set_small_glyph_heights(True)`` before searching and defining annot rectangles.
